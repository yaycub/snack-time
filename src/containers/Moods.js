import React, { Component } from 'react';
import Controls from '../components/controls/Controls';
import Face from '../components/face/Face';

const actions = [
  { name: 'DRINK_COFFEE', text: 'Drink Coffee', stateName: 'coffees' },
  { name: 'EAT_SNACK', text: 'Snack', stateName: 'snacks' },
  { name: 'TAKE_NAP', text: 'Nap', stateName: 'naps' },
  { name: 'STUDY', text: 'Study', stateName: 'studies' },
];

export const isTired = state => state.coffees < 1 && state.naps < 1;
export const isHyper = state => state.coffees > 3;
export const isEducated = state => state.studies > 2;
export const isHungry = state => state.snacks < 1;

export const getFace = state => {
  if(isTired(state) && isHungry(state)) return 'ðŸ˜ ';
  if(isHyper(state) && isHungry(state)) return 'ðŸ˜±';
  if(isTired(state)) return 'ðŸ˜´';
  if(isHyper(state)) return 'ðŸ™€';
  if(isEducated(state)) return 'ðŸ˜²';
  if(isHungry(state)) return 'ðŸ˜¡';

  return 'ðŸ˜€';
};

export default class Moods extends Component {
  state = {
    coffees: 0,
    snacks: 0,
    naps: 0,
    studies: 0
  }

  handleSelection = name => {
    switch(name) {
      case 'DRINK_COFFEE':
        this.setState(state => ({ coffees: state.coffees + 1 }));
        break;
      case 'EAT_SNACK':
        this.setState(state => ({ snacks: state.snacks + 1 }));
        break;
      case 'TAKE_NAP':
        this.setState(state => ({ naps: state.naps + 1 }));
        break;
      case 'STUDY':
        this.setState(state => ({ studies: state.studies + 1 }));
        break;
      default:
        console.log(`unhandled name: ${name}`);
    }
  }

  render() {
    const face = getFace(this.state);
    const controlActions = actions.map(action => ({
      ...action,
      count: this.state[action.stateName]
    }));

    return (
      <>
        <Controls actions={controlActions} handleSelection={this.handleSelection}/>
        <Face emoji={face} />
      </>
    );
  }
}
