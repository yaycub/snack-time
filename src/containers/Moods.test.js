import React from 'react';
import { shallow } from 'enzyme';
import Moods, { isTired, isHyper, isEducated, isHungry, getFace } from './Moods';

describe('Moods', () => {
  describe('mood selectors', () => {
    it('knows if it is tired', () => {
      expect(isTired({ coffees: 0, naps: 0 })).toBeTruthy();
    });

    it('knows if it is not tired', () => {
      expect(isTired({ coffees: 1, naps: 0 })).toBeFalsy();
    });

    it('knows if it is hyper', () => {
      expect(isHyper({ coffees: 4 })).toBeTruthy();
    });

    it('knows if it is not hyper', () => {
      expect(isHyper({ coffees: 3 })).toBeFalsy();
    });

    it('knows if it is educated', () => {
      expect(isEducated({ studies: 3 })).toBeTruthy();
    });

    it('knows if it is not educated', () => {
      expect(isEducated({ studies: 2 })).toBeFalsy();
    });

    it('knows if it is hungry', () => {
      expect(isHungry({ snacks: 0 })).toBeTruthy();
    });

    it('knows if it is not hungry', () => {
      expect(isHungry({ studies: 2 })).toBeFalsy();
    });

    it('is very angry if tired and hungry', () => {
      expect(getFace({ coffees: 0, naps: 0, snacks: 0 })).toEqual('ðŸ¤¬');
    });

    it('is sick if hyper and hungry', () => {
      expect(getFace({ coffees: 4, snacks: 0 })).toEqual('ðŸ¤®');
    });

    it('is sleeping if tired', () => {
      expect(getFace({ coffees: 4, snacks: 0 })).toEqual('ðŸ¤®');
    });

    it('is crazy if hyper', () => {
      expect(getFace({ coffees: 4, naps: 0, snacks: 0, studies: 0 })).toEqual('ðŸ¤®');
    });

    it('is mindblown if educated', () => {
      expect(getFace({ coffees: 1, naps: 0, snacks: 1, studies: 3 })).toEqual('ðŸ¤¯');
    });

    it('is angry if hungry', () => {
      expect(getFace({ coffees: 1, naps: 0, snacks: 0, studies: 0 })).toEqual('ðŸ˜¡');
    });

    it('is happy if not tired, hungry, educated, or hungry', () => {
      expect(getFace({ coffees: 1, naps: 0, snacks: 1, studies: 0 })).toEqual('ðŸ˜€');
    });
  });

  describe('container', () => {
    it('matches a snapshot', () => {
      const wrapper = shallow(<Moods />);
      expect(wrapper).toMatchSnapshot();
    });

    it('updates state on DRINK_COFFEE selection', () => {
      const wrapper = shallow(<Moods />);
      wrapper.instance().handleSelection('DRINK_COFFEE');

      expect(wrapper.state('coffees')).toEqual(1);
    });

    it('updates state on TAKE_NAP selection', () => {
      const wrapper = shallow(<Moods />);
      wrapper.instance().handleSelection('TAKE_NAP');

      expect(wrapper.state('naps')).toEqual(1);
    });

    it('updates state on EAT_SNACK selection', () => {
      const wrapper = shallow(<Moods />);
      wrapper.instance().handleSelection('EAT_SNACK');

      expect(wrapper.state('snacks')).toEqual(1);
    });

    it('updates state on STUDY selection', () => {
      const wrapper = shallow(<Moods />);
      wrapper.instance().handleSelection('STUDY');

      expect(wrapper.state('studies')).toEqual(1);
    });
  });
});
