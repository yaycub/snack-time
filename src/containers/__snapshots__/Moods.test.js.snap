// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Moods container matches a snapshot 1`] = `
<Fragment>
  <Controls
    actions={
      Array [
        Object {
          "name": "DRINK_COFFEE",
          "text": "Drink Coffee",
        },
        Object {
          "name": "EAT_SNACK",
          "text": "Snack",
        },
        Object {
          "name": "TAKE_NAP",
          "text": "Nap",
        },
        Object {
          "name": "STUDY",
          "text": "Study",
        },
      ]
    }
    handleSelection={[Function]}
  />
  <Face
    emoji="ðŸ¤¬"
  />
</Fragment>
`;
